cli/commands/stack/action.go

```diff
@@ -0,0 +1,228 @@
+package stack
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/gruntwork-io/terragrunt/config"
+	getter "github.com/hashicorp/go-getter"
+
+	"github.com/gruntwork-io/terragrunt/internal/errors"
+	"github.com/gruntwork-io/terragrunt/options"
+)
+
+const (
+	generate         = "generate"
+	stackCacheDir    = ".terragrunt-stack"
+	defaultStackFile = "terragrunt.stack.hcl"
+	dirPerm          = 0755
+)
+
+func Run(ctx context.Context, opts *options.TerragruntOptions, subCommand string) error {
+	if subCommand == "" {
+		return errors.New("No command specified")
+	}
+
+	if subCommand == generate {
+		return generateStack(ctx, opts)
+	}
+
+	return nil
+}
+
+func generateStack(ctx context.Context, opts *options.TerragruntOptions) error {
+	opts.TerragrungStackConfigPath = filepath.Join(opts.WorkingDir, defaultStackFile)
+	stackFile, err := config.ReadStackConfigFile(ctx, opts)
+
+	if err != nil {
+		return errors.New(err)
+	}
+
+	if err := processStackFile(ctx, opts, stackFile); err != nil {
+		return errors.New(err)
+	}
+
+	return nil
+}
+func processStackFile(ctx context.Context, opts *options.TerragruntOptions, stackFile *config.StackConfigFile) error {
+	baseDir := filepath.Join(opts.WorkingDir, stackCacheDir)
+	if err := os.MkdirAll(baseDir, dirPerm); err != nil {
+		return errors.New(fmt.Errorf("failed to create base directory: %w", err))
+	}
+
+	for _, unit := range stackFile.Units {
+		destPath := filepath.Join(baseDir, unit.Path)
+		dest, err := filepath.Abs(destPath)
+
+		if err != nil {
+			return errors.New(fmt.Errorf("failed to get absolute path for destination '%s': %w", dest, err))
+		}
+
+		src := unit.Source
+		// set absolute path for source if it's not an absolute path or URL
+		if !filepath.IsAbs(unit.Source) && !isURL(unit.Source) {
+			src = filepath.Join(opts.WorkingDir, unit.Source)
+			src, err = filepath.Abs(src)
+
+			if err != nil {
+				opts.Logger.Warnf("failed to get absolute path for source '%s': %v", unit.Source, err)
+				src = unit.Source
+			}
+		}
+
+		opts.Logger.Infof("Processing unit: %s (%s) to %s", unit.Name, src, dest)
+
+		client := &getter.Client{
+			Src:             src,
+			Dst:             dest,
+			Mode:            getter.ClientModeAny,
+			Dir:             true,
+			DisableSymlinks: true,
+			Options: []getter.ClientOption{
+				getter.WithInsecure(),
+				getter.WithContext(ctx),
+			},
+		}
+
+		// setting custom getters
+		client.Getters = map[string]getter.Getter{}
+
+		for getterName, getterValue := range getter.Getters {
+			// setting custom getter for file to not use symlinks
+			if getterName == "file" {
+				client.Getters[getterName] = &StacksFileProvider{}
+			} else {
+				client.Getters[getterName] = getterValue
+			}
+		}
+
+		if err := client.Get(); err != nil {
+			return errors.New(err)
+		}
+	}
+
+	return nil
+}
+
+func isURL(str string) bool {
```

Comment:
Does this mean that `github.com/...` URLs won't be supported?

Reply:
It might be better to see if the source starts with `.` or `/` to indicate that it's a relative or absolute file path respectively.

Reply:
In case of `github.com/` you don't know if it is local directory or remote... added workaround to check for `//`

Reply:
I was saying that we can require that local sources always start with `./` for a relative path or `/` for an absolute path.

I'm not sure a user would ever want to do this, but could they define their stack definition at the root of a repository?

e.g.

`github.com/acme/my-favorite-stack`